### PR TITLE
[DDP] Fix static_graph=True + no_sync() gradient accumulation regression

### PR DESCRIPTION
**Pull request by kasper0406**
DDP(static_graph=True) with gradient accumulation via no_sync() crashes on the first backward with an expect_autograd_hooks_ INTERNAL ASSERT, and once that is unblocked it silently stops all-reducing gradients from the second iteration on. Both are regressions from #103487, which re-keyed the static-graph "first iteration" check from num_iterations_ (advances only on syncing forwards) to num_bwd_calls_ (advances on every backward). Reported in huggingface/accelerate#3679.

Under no_sync() require_backward_grad_sync is False, so _post_forward skips prepare_for_backward() and expect_autograd_hooks_ stays false for the whole accumulation window. The crash: _DDPSink.backward still enqueued the one-shot static-graph delay_all_reduce, whose finalize_backward() asserts expect_autograd_hooks_; it now only enqueues on a syncing backward. The wrong gradients: with num_bwd_calls_ stuck at 1, static_graph_first_iteration() stays true, so every micro-batch backward incremented numGradHooksTriggeredMap_, inflating the per-parameter counts until the per-iteration counter could no longer reach zero and parameters were never marked ready; the increment is now gated on expect_autograd_hooks_ so only the syncing backward records it. This preserves https://github.com/pytorch/pytorch/pull/103487's multiple-forward support, whose single backward syncs.

Test Plan:
New test_static_graph_no_sync_gradient_accumulation runs static_graph + no_sync accumulation for several iterations, checking grads against an all-reduced local reference, with distinct per-rank data so a missing all-reduce is visible. Without the Python fix it asserts on iteration 1; without the C++ fix it fails the grad check from iteration 2.

```bash
touch /tmp/barrier
TEMP_DIR=/tmp BACKEND=gloo WORLD_SIZE=2 python \
    test/distributed/test_distributed_spawn.py -v \
    -k test_static_graph_no_sync_gradient_accumulation
```

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @kapilsh

### CODE REVIEW

**Comment by kasper0406** [comment]
@weifengpy are you able to give this a look? The issue is preventing me from doing DDP training runs with static graph no_sync gradient accumulation. The failing CI check looks to be a fluke.
